// @cyclonedx/cyclonedx-npm/dist/_helpers.js
"use strict";
/*!
This file is part of CycloneDX generator for NPM projects.

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

   http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.

SPDX-License-Identifier: Apache-2.0
Copyright (c) OWASP Foundation. All Rights Reserved.
*/
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.setDifference = void 0;
exports.loadJsonFile = loadJsonFile;
exports.writeAllSync = writeAllSync;
exports.isString = isString;
exports.tryRemoveSecretsFromUrl = tryRemoveSecretsFromUrl;
exports.versionTuple = versionTuple;
exports.versionCompare = versionCompare;
exports.iterableMap = iterableMap;
exports.iterableFilter = iterableFilter;
exports.normalizePackageManifest = normalizePackageManifest;
const node_fs_1 = require("node:fs");
const normalize_package_data_1 = __importDefault(require("normalize-package-data"));
function loadJsonFile(path) {
    return JSON.parse((0, node_fs_1.readFileSync)(path, 'utf8'));
}
async function writeAllSync(fd, data) {
    const b = Buffer.from(data);
    const l = b.byteLength;
    let w = 0;
    while (w < l) {
        try {
            w += (0, node_fs_1.writeSync)(fd, b, w);
        }
        catch (error) {
            if (error.code !== 'EAGAIN') {
                throw error;
            }
            await new Promise((resolve) => { setTimeout(resolve, 100); });
        }
    }
    return w;
}
function isString(v) {
    return typeof v === 'string';
}
function tryRemoveSecretsFromUrl(url) {
    try {
        const u = new URL(url);
        u.password = '';
        return u.toString();
    }
    catch {
        return url;
    }
}
function versionTuple(value) {
    return Object.freeze(value.split('.').map(Number));
}
function versionCompare(a, b) {
    for (let i = 0, l = Math.max(a.length, b.length); i < l; ++i) {
        const ai = a[i] || 0;
        const bi = b[i] || 0;
        if (ai < bi) {
            return -1;
        }
        if (ai > bi) {
            return 1;
        }
    }
    return 0;
}
function* iterableMap(iter, func) {
    for (const item of iter) {
        yield func(item);
    }
}
function* iterableFilter(iter, func) {
    for (const item of iter) {
        if (func(item)) {
            yield item;
        }
    }
}
exports.setDifference = typeof Set.prototype.difference === 'function'
    ? (s1, s2) => s1.difference(s2)
    : (s1, s2) => new Set(iterableFilter(s1, (i) => !s2.has(i)));
function normalizePackageManifest(data) {
    const oVersion = data.version;
    (0, normalize_package_data_1.default)(data);
    if (isString(oVersion)) {
        data.version = oVersion.trim();
    }
}
